File: 3-HARD/Python/3-suffix-array-construction/proposed-solution/main.py
# Challenge: Construct a suffix array (and optionally an LCP array) for a given string,
# enabling very efficient pattern searches. This is an advanced challenge in string processing.
# Use object-oriented programming and follow the DRY principle.

from suffix_array import SuffixArray

def main():
    text = "banana"
    sa = SuffixArray(text)
    suffix_arr = sa.build_suffix_array()
    lcp_arr = sa.build_lcp_array()
    print(f"Suffix Array: {suffix_arr}")
    print(f"LCP Array: {lcp_arr}")

if __name__ == "__main__":
    main()

File: 3-HARD/JavaScript/3-suffix-array-construction/proposed-solution/main.js
/*
Challenge:
Construct a suffix array and LCP array for a given string to enable efficient pattern searches.

This solution follows DRY principles and is implemented in JavaScript.
*/

import SuffixArray from './SuffixArray.js';

function main() {
  const text = "banana";
  const suffixArrayObj = new SuffixArray(text);

  const suffixArray = suffixArrayObj.buildSuffixArray();
  console.log("Suffix Array:", suffixArray);

  const lcpArray = suffixArrayObj.buildLCPArray();
  console.log("LCP Array:", lcpArray);
}

main();
